module/os/freebsd/zfs/zfs_vfsops.c

```diff
@@ -1310,6 +1310,18 @@ zfs_domount(vfs_t *vfsp, char *osname)
 
 	if (!zfsvfs->z_issnap)
 		zfsctl_create(zfsvfs);
+
+#ifdef __FreeBSD__
+	if (error == 0) {
+		/* zone dataset visiblity was checked before in zfs_mount() */
+		struct prison *pr = curthread->td_ucred->cr_prison;
+		if (pr != &prison0) {
+			zfsvfs->z_os->os_dsl_dataset->ds_jailname =
+			    kmem_strdup(pr->pr_name);
```

Comment:
Again, I am not closely familiar with the zone/jail code in ZFS, but I am not sure the delegation should necessarily be tied to a mounting.  I suppose some dataset may not be mounted, but be controlled by the jail.  Am I wrong?